authorization/client/ownerCache.go

```diff
@@ -0,0 +1,355 @@
+package client
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	nats "github.com/nats-io/nats.go"
+	"github.com/plgd-dev/cloud/authorization/events"
+	pbAS "github.com/plgd-dev/cloud/authorization/pb"
+	"github.com/plgd-dev/cloud/pkg/log"
+	kitNetGrpc "github.com/plgd-dev/cloud/pkg/net/grpc"
+	"github.com/plgd-dev/cloud/pkg/strings"
+	"github.com/plgd-dev/cloud/resource-aggregate/cqrs/utils"
+	kitSync "github.com/plgd-dev/kit/sync"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type ownerSubject struct {
+	handlers         map[uint64]func(e *events.Event)
+	subscription     *nats.Subscription
+	devices          strings.SortedSlice
+	validUntil       time.Time
+	devicesWasSynced bool
+	sync.Mutex
+}
+
+func newOwnerSubject(validUntil time.Time) *ownerSubject {
+	return &ownerSubject{
+		handlers:   make(map[uint64]func(e *events.Event)),
+		devices:    make(strings.SortedSlice, 0, 16),
+		validUntil: validUntil,
+	}
+}
+
+func (d *ownerSubject) Handle(msg *nats.Msg) error {
+	var e events.Event
+	err := utils.Unmarshal(msg.Data, &e)
+	if err != nil {
+		return err
+	}
+	d.Lock()
+	if d.devicesWasSynced {
+		for _, deviceID := range e.GetDevicesRegistered().GetDeviceIds() {
+			d.devices = strings.Insert(d.devices, deviceID)
+		}
+		for _, deviceID := range e.GetDevicesUnregistered().GetDeviceIds() {
+			d.devices = strings.Remove(d.devices, deviceID)
+		}
+	}
+	handlers := make(map[uint64]func(e *events.Event))
+	for key, h := range d.handlers {
+		handlers[key] = h
+	}
+	d.Unlock()
+	for _, h := range handlers {
+		h(&e)
+	}
+
+	return nil
+}
+
+func (d *ownerSubject) AddHandlerLocked(id uint64, h func(e *events.Event)) bool {
+	if _, ok := d.handlers[id]; !ok {
+		d.handlers[id] = h
+		return true
+	}
+	return false
+}
+
+func (d *ownerSubject) RemoveHandlerLocked(v uint64) {
+	delete(d.handlers, v)
+}
+
+func (d *ownerSubject) updateDevicesLocked(deviceIDs []string) (added []string, removed []string) {
+	added = make([]string, 0, 8)
+	removed = make([]string, 0, 8)
+	deviceIDs = strings.MakeSortedSlice(deviceIDs)
+	var j int
+	for i := range deviceIDs {
+		if j >= len(d.devices) || deviceIDs[i] != d.devices[j] {
```

Comment:
is devicesIds guaranteed to contains everything in d.devices?

because for example:

deviceIDs = ["dev2", "dev3", "dev4"]
d.devices = ["dev1", "dev2", "dev3", "dev4"]

===============
added = ["dev2", "dev3", "dev4"]

which doesn't seem right?